homeassistant/components/remote/harmony.py

```diff
@@ -0,0 +1,176 @@
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_PORT): cv.string,
+    vol.Required(ATTR_ACTIVITY, default='Not_Set'): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup Harmony platform."""
+    import pyharmony
+    import urllib.parse
+    name = config.get(CONF_NAME)
+    user = config.get(CONF_USERNAME)
+    passw = config.get(CONF_PASSWORD)
+    _LOGGER.info('Loading Harmony platform: ' + name)
+
+    harmony_conf_file = hass.config.path('harmony_' + slugify(name) + '.conf')
+
+    try:
+        _LOGGER.debug('calling pyharmony.ha_get_token with username: ' +
+                      user + ' and password: ' + passw)
+        token = urllib.parse.quote_plus(pyharmony.ha_get_token(user, passw))
+    except ValueError as err:
+        _LOGGER.critical(err.args[0] + ' for remote: ' + name)
+        return False
+
+    _LOGGER.debug('received token: ' + token)
+    add_devices([HarmonyRemote(config.get(CONF_NAME),
+                               config.get(CONF_USERNAME),
+                               config.get(CONF_PASSWORD),
+                               config.get(CONF_HOST),
+                               config.get(CONF_PORT),
+                               config.get(ATTR_ACTIVITY),
+                               harmony_conf_file,
+                               token)])
+    return True
+
+
+class HarmonyRemote(remote.RemoteDevice):
+    """Remote representation used to control a Harmony device."""
+
+    def __init__(self, name, username, pw, host, port, activity, path, token):
+        """Initialize HarmonyRemote class."""
+        import pyharmony
+        from pathlib import Path
+        _LOGGER.debug('HarmonyRemote device init started for: ' + name)
+        self._name = name
+        self._email = username
+        self._password = pw
+        self._ip = host
+        self._port = port
+        self._state = None
+        self._current_activity = None
+        self._default_activity = activity
+        self._token = token
+        self._config_path = path
+        _LOGGER.debug('retrieving hub config using token: ' + token)
+        self._config = pyharmony.ha_get_config(self.token, host, port)
+        if not Path(self._config_path).is_file():
+            _LOGGER.debug('writing hub configuration to file: ' + path)
+            pyharmony.ha_get_config_file(self._config, self._config_path)
+
+    @property
+    def name(self):
+        """Return the Harmony device's name."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the Harmony device."""
+        return self._state
+
+    @property
+    def token(self):
+        """Return the token of the Harmony device."""
+        return self._token
+
+    @property
+    def config(self):
+        """Return the configuration of the Harmony device."""
+        return self._config
+
+    @property
+    def state_attributes(self):

```

Comment:
Platforms should not overwrite `state_attributes` but implement `device_state_attributes` instead.